cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java

```diff
@@ -0,0 +1,253 @@
+                                        " but got " + context.getClass().getName());
+    }
+
+    started = true;
+  }
+
+  @Override
+  public void doAppend(ILoggingEvent eventObject) throws LogbackException {
+    try {
+      // logic from AppenderBase
+      if (!this.started) {
+        LOG.warn("Attempted to append to non started appender {}", this.getName());
+        return;
+      }
+
+      // logic from AppenderBase
+      if (getFilterChainDecision(eventObject) == FilterReply.DENY) {
+        return;
+      }
+
+      String namespaceId = eventObject.getMDCPropertyMap().get(LocationManager.TAG_NAMESPACE_ID);
+
+      if (namespaceId != null && !namespaceId.equals(NamespaceId.SYSTEM.getNamespace())) {
+        LocationIdentifier logLocationIdentifier = locationManager.getLocationIdentifier(eventObject
+                                                                                           .getMDCPropertyMap());
+        rollover(logLocationIdentifier, eventObject);
+        OutputStream locationOutputStream = locationManager.getLocationOutputStream(logLocationIdentifier, filePath);
+        setOutputStream(locationOutputStream);
+        writeOut(eventObject);
+      }
+    } catch (IllegalArgumentException iae) {
+      // this shouldn't happen
+      LOG.error("Unrecognized context ", iae);
+    } catch (IOException ioe) {
+      throw new LogbackException("Exception while appending event. ", ioe);
+    }
+  }
+
+  private void rollover(final LocationIdentifier logLocationIdentifier, ILoggingEvent logEvent) {
+    if (!locationManager.getActiveLocations().containsKey(logLocationIdentifier)) {
+      LOG.info("Decided not to roll over for first event");
+      return;
+    }
+    final LocationOutputStream locationOutputStream = locationManager.getActiveLocations().get(logLocationIdentifier);
+
+    if (triggeringPolicy instanceof LocationTriggeringPolicy) {
+      // no need to type cast on every event
+      if (locationTriggeringPolicy == null) {
+        locationTriggeringPolicy = ((LocationTriggeringPolicy) triggeringPolicy);
+      }
+
+      locationTriggeringPolicy.setLocation(locationOutputStream.getLocation());
+      // set number of bytes written to locationOutputStream, we need to do this because HDFS does not provide
+      // correct size of the file
+      locationTriggeringPolicy.setActiveLocationSize(locationOutputStream.getNumOfBytes());
+
+      if (locationTriggeringPolicy.isTriggeringEvent(logEvent)) {
+        try {
+          if (rollingPolicy instanceof LocationRollingPolicy) {
+            // no need to type cast on every event
+            if (locationRollingPolicy == null) {
+              locationRollingPolicy = ((LocationRollingPolicy) rollingPolicy);
+            }
+
+            locationRollingPolicy.setLocation(locationOutputStream.getLocation(), new Closeable() {
+              @Override
+              public void close() throws IOException {
+                locationOutputStream.close();
```

Comment:
If close fail, it means the outputstream could be in some bad state that may not be recoverable from the program space. So it's better not to keep writing to it. However, it does create a problem that the next event that come will try to open a new file with the same file name (assuming we don't rename or the rename failed as well), which is not good. I think the best sequence is this:

I think the best sequence is something like this:

1. Remove the stream from the map
2. Attempt to close the file. If it fails, remember the output stream and throw exception. The framework will log a warn and proceed.
3. If step 2 succeeded, try to rename the file. If rename failed, remember the location and throw exception. The framework will log a warn and proceed.
4. Next time when `append()` is called, if either step 2 or 3 was a failure, retry the close/rename based on the remembered stream/location before appending (hence creating new file). If this fail, throw exception. The framework will handle it by sending the same event again.
5. If step 4 cannot proceed, the framework will keep repeating step 4. It might result in filling up the in memory sort buffer, hence stop polling from Kafka. I think that's ok as that's the back pressure mechanism in the framework and it still guarantees there is no loss of data.

Note that the framework only has this mechanism in distributed mode, as we have Kafka to act as the backing buffer. In local mode, everything happen in memory. If there is a failure to Append, the event will simply be ignore (the chance that it fails in local mode is small anyway).